### DISCUSSION TITLE
Examples!

### DISCUSSION THREAD

**Discussion by brainkim** [bikeshaving/crank#38] [General]
I see your comments and appreciate all the design feedback, but I must admit, writing long technical responses can be both intellectually and emotionally draining. Know that I see you and I will get back to you ASAP about my thoughts on all of this stuff.

One thing that I love more than anything is seeing examples! Have you made Snake or Tetris? Let me know in this issue 🙆‍♀️

**Comment by shirakaba** [comment]
I've started making a spinning Web Browser in Crank Native:

https://play.nativescript.org/?template=play-react&id=LB4qRS&v=4

It accesses native iOS methods to inject some JS into the webpage to make things, uh, spin.

![IMG_9619](https://user-images.githubusercontent.com/14055146/79693167-05dc8c80-8261-11ea-8d9e-31453d7cafbb.PNG)

**Comment by shirakaba** [comment]
Built an iOS SpriteKit game launcher in Crank Native, which confirms that the NativeScript Playground supports generators!

https://play.nativescript.org/?template=play-react&id=ptkGyA&v=8

![IMG_9620](https://user-images.githubusercontent.com/14055146/79693871-0bd46c80-8265-11ea-9e06-f1a53588ff27.PNG)

![IMG_9624](https://user-images.githubusercontent.com/14055146/79693876-10992080-8265-11ea-9c83-a8a90ebd239f.PNG)

**Comment by ryhinchey** [comment]
I came across this on Twitter:


https://codesandbox.io/embed/crankjs-final-form-jk3q9?file=/src/index.js&codemirror=1


https://twitter.com/erikras/status/1251584285192138753?s=21

**Comment by btheado** [comment]
I copied the code examples from https://crank.js.org/guides/components to https://observablehq.com/@btheado/crank-js and translated them to use [htm](https://github.com/developit/htm). All the examples are live-editable and runnable.

When I first read about crank.js the other day, I was reminded of the way observablehq uses generators and async generators. I haven't yet compare/contrasted them, so I don't know how similar they are. I thought I'd play with crank on observable with the eventual goal of finding out. The above page is the result.

Update: I also added the code examples from https://crank.js.org/guides/async-components.

In addition, I found this other example also on observable: https://observablehq.com/@radames/hello-crank-js

**Comment by ryhinchey** [comment]
I just put this together using the undocumented `this.set` and `this.get` (kind of like React contexts) https://codesandbox.io/s/lucid-bash-nh846?file=/src/index.js

```jsx
/** @jsx createElement */
import { createElement, Fragment } from "@bikeshaving/crank";
import { renderer } from "@bikeshaving/crank/dom";

function Avatar() {
  const user = this.get("user");
  return <img alt={`${user.first_name} ${user.last_name}`} src={user.avatar} />;
}

function User() {
  const user = this.get("user");

  return (
    user && (
      <div>
        <p>{user.first_name}</p>
        <Avatar image={user.avatar} />
      </div>
    )
  );
}

async function* App() {
  try {
    let value = "";

    const onChange = e => {
      value = e.target.value;
      this.refresh();
    };

    const onSubmit = async (e) => {
      e.preventDefault();
      const response = await fetch(`https://reqres.in/api/users/${value}`);
      const user = await response.json();
      this.set("user", user.data);
      this.refresh();
    };

    for await (let _ of this) {
      yield (
        <Fragment>
          <form onsubmit={onSubmit}>
            <label for="userSelector">Select a User Number (1 - 10)</label>
            <input
              value={value}
              onchange={onChange}
              id="userSelector"
              type="number"
              max="10"
              min="1"
            />
            <button type="submit">Get new User</button>
          </form>
          <p>Selected Number: {value}</p>
          <User />
        </Fragment>
      );
    }
  } catch (error) {
    return <h1>error getting data </h1>;
  }
}

renderer.render(<App />, document.body);

```

The `App` component is an async generator and I'm using `for await (let _ of this)` for my UI to respond to changes.  I'm not using any props and am getting an eslint warning for unused variable `_`.    @brainkim What's the suggestion here?  I'm pretty sure I need to use the `for await of`.

Also - `this.set` and `this.get` are great to have in the framework (Although I think we might want to move them to something like `this.proisions.get` or `this.context.get`)

**Reply by lazeebee** [reply]
This will work around the eslint warning
```js
async function* App() {
  try {
    let value = "";

    const onChange = e => {
      value = e.target.value;
      this.refresh();
    };
    const onSubmit = async e => {
      e.preventDefault();
      const response = await fetch(`https://reqres.in/api/users/${value}`);
      const user = await response.json();
      this.set("user", user.data);
      this.refresh();
    };

    const props = this[Symbol.asyncIterator]();

    while (true) {
      await props.next();
      yield (
        <Fragment>
          <form onsubmit={onSubmit}>
            <label for="userSelector">Select a User Number (1 - 10)</label>
            <input
              value={value}
              onchange={onChange}
              id="userSelector"
              type="number"
              max="10"
              min="1"
            />
            <button type="submit">Get new User</button>
          </form>
          <p>Selected Number: {value}</p>
          <User />
        </Fragment>
      );
    }
  } catch (error) {
    return <h1>error getting data </h1>;
  }
}
```

**Reply by brainkim** [reply]
That‘s definitely a way to do it @lazeebee. Honestly though I’m always hesitant to put `while (true)` loops in async generator functions because if you forget to await props or some other async thing you will end up wondering why your laptop fan is running after you finish developing and call `top` to find out there’s two node processes jointly consuming 96% of CPU. I think it’s just better to make  `for await (const _ of this)` a pattern and change our tooling to allow it.

**Reply by ryhinchey** [reply]
Agreed we’ll have to build an Eslint plugin that allows this.  It could also warn you of using while true in async generators, etc

**Reply by brainkim** [reply]
I’ve just discovered that you can do `for ({} of this)`. Could this be the solution?

**Comment by brainkim** [comment]
Here‘s the XState calculator example in Crank: https://codesandbox.io/s/xstate-calculator-with-crankjs-u9r0f?file=/src/index.tsx

I see that there could probably be some kinda of plugin created which refreshes on transition automatically I just don’t know what it should look like yet. In the meantime it just works with a couple helper functions and an explicit refresh call.

**Comment by brainkim** [comment]
https://github.com/brainkim/crank-three-boxes

Three.js boxes demo. Couldn’t get it to work in Code Sandbox for some reason. If you are interested in writing a Three.js renderer for Crank reach out and I will guide you through the forbidden/unstable renderer APIs.

**Comment by jonrsawyer** [comment]
Just playing around with Material Design components (actually, "component", at the moment; only a Button so far) with Crank:

https://github.com/jonrsawyer/crank-material

**Reply by brainkim** [reply]
This is cool! Let me know if you have any feedback or questions and I will gladly help out! You can use async generator components here but we will have a sync way to access the underlying DOM nodes soon. There is some work going on for some limited plugin-like functionality so you may soon be able to abstract the MDCRipple out of the class.

**Reply by jonrsawyer** [reply]
The MDCRipple mechanism requires the <button> tag to be inserted into the document; calling MDCRipple after the yield seemed to be a painless way to do this. But maybe plugins will be even better...

**Reply by jonrsawyer** [reply]
> Let me know if you have any feedback or questions and I will gladly help out!

Well, since you offered...

I've pushed a new commit of my https://github.com/jonrsawyer/crank-material repo with a problem. I suspect the problem is that the Crank.js diffing algorithm may not be handling SVG elements properly. If you "npm start" the repo, you'll see two checkboxes that appear differently when checked. The first checkbox is created from a Crank component; the second is statically declared in the index.html file. The SVG path for the checkmark does not appear in the former.

I'm kinda out of my depth here, but https://stackoverflow.com/questions/10546135/appending-path-child-within-svg-using-javascript/10546700 seems like it may apply here.

Or more likely, I'm doing something stupid...

**Reply by brainkim** [reply]
I see this and promise to triage by tomorrow noon EST. I assure you I am the stupid one here, most likely.

**Reply by jonrsawyer** [reply]
I appreciate your "I see this" comment. FYI, I'm a real newbie at Github etiquette; was I supposed to do something to explicitly bring this to your attention?

And another thing...

I'm working on another set of components using async generators that consist of a parent component and a set of child components (Material design's ChipSet and Chip). In the parent component, I'm using the `const div = yield ( <jsx declarations> )` to get hold of the created DOM element. However, in this case, in the parent component, `div` is a Promise, not an Element. Is this to be expected from Crank's `createElement` call? Should I always `await` what `yield` returns to ensure I have an Element? Perhaps that should be a recommended idiom? FWIW, I'm not sure what I'm doing differently in these components to cause the appearance of a Promise, other than the fact that I'm creating components within components.

Finally: is there a better forum for these sorts of inquiries than continuing to extend the comments on an item in this "Examples!" discussion?

**Reply by brainkim** [reply]
> I appreciate your "I see this" comment. FYI, I'm a real newbie at Github etiquette; was I supposed to do something to explicitly bring this to your attention?

I see everything 👀 Working on adding SVG support in Crank right now. You can usually get the attention of people by \@-ing them as in @brainkim. People will like/dislike this according to their personalities I guess.

> I'm working on another set of components using async generators that consist of a parent component and a set of child components (Material design's ChipSet and Chip). In the parent component, I'm using the const div = yield ( <jsx declarations> ) to get hold of the created DOM element. However, in this case, in the parent component, div is a Promise, not an Element. Is this to be expected from Crank's createElement call? Should I always await what yield returns to ensure I have an Element? Perhaps that should be a recommended idiom? FWIW, I'm not sure what I'm doing differently in these components to cause the appearance of a Promise, other than the fact that I'm creating components within components.

Yes that is a consequence of yielding an element tree with an async component. The whole tree doesn’t commit until the async component has fulfilled, so it resumes with a promise of the result rather than the result itself. If you know none of the subtree is async though, you don’t have to add an extra await, which can be convenient. You typically know what was just yielded unless you’re dynamically rendering children.

> Finally: is there a better forum for these sorts of inquiries than continuing to extend the comments on an item in this "Examples!" discussion?

You can create more discussions! And if you think there’s something actionable like a bug you can open an issue. As far as other places to discuss this stuff. It doesn’t really exist yet. We’ll play it by ear. Welcome to GitHub!

**Reply by brainkim** [reply]
@jonrsawyer SVG support has been added in 0.1.4
